### PR TITLE
.ovpn12 files

### PR DESCRIPTION
**Pull request by IcedComputer**
Added new step to create an .ovpn12 file that can be stored on iOS keychain
This step is more secure method and does not require the end-user to keep entering passwords, or storing the client private cert where it can be easily tampered based on documentation located:
 https://openvpn.net/faq/how-do-i-use-a-client-certificate-and-private-key-from-the-ios-keychain/

Someone can improve upon this by adding a parameter (possibly -i|--iOS) and then generating the original .ovpn file to not contain the client private certificate.

### CODE REVIEW

**Comment by 4s3ti** [comment]
Will let other most active users comment on this as I am not at all familiar with IOS devices. 

If their feedback is positive then ill gladly merge it. 

@orazioedoardo @devZer0 @Giraffe1966  any takes on this?

**Comment by orazioedoardo** [comment]
Yeah, I see that this is more secure because the private parts of the config are not in the .ovpn file, though it's a bit less convenient since the user need to copy two files, both the .ovpn and the .ovpn12 and instruct the user on how to use them. Also your script doesn't do the whole job:
- the .ovpn12 file is root owned in the ovpn folder so you should chown and chmod it to allow the user to read it.
- you need to strip cert and key sections (or just don't add them) in the .ovpn otherwise you defeat initial purpose.
- probably you should use expect to type the export password and the private key password non-interactively.

By the way I wouldn't make this the default, I would add a one time dialog after the first client creation with some text like "TIP: If you have an iOS device you can pass -i|--iOS to [...]"

**Comment by Giraffe1966** [comment]
No need to use expect for the password, you can use the `-passin env:PASSWD` and `-passout env:PASSWD` options to openssl to do this non-interactively. Also, don't use sudo for openssl as this is run as root and this would break on systems without sudo installed.

**Comment by IcedComputer** [comment]
Thanks for the feedback.  I will be sure to incorporate it.


> * you need to strip cert and key sections (or just don't add them) in the .ovpn otherwise you defeat initial purpose.

> By the way I wouldn't make this the default, I would add a one time dialog after the first client creation with some text like "TIP: If you have an iOS device you can pass -i|--iOS to [...]"

I am a bit new at this, so trying to figure out how to use the -i|--iOS filter correctly to create this and strip out the cert and key sections.  I will do some testing and upload something that works!

**Comment by IcedComputer** [comment]
Updated.

**Comment by 4s3ti** [comment]
@orazioedoardo , @Giraffe1966  what is your take on this? is it good to be merged? 

thanks a lot for keeping the boat moving! =)

File: scripts/makeOVPN.sh [hunk 1]
```diff
@@ -13,14 +13,15 @@ INSTALL_USER=$(cat /etc/pivpn/INSTALL_USER)
...
-    echo "::: Usage: pivpn <-a|add> [-n|--name <arg>] [-p|--password <arg>]|[nopass] [-d|--days <number>] [-h|--help]"
+    echo "::: Usage: pivpn <-a|add> [-n|--name <arg>] [-p|--password <arg>]|[nopass] [-d|--days <number>] [-i|--iOS] [-h|--help]"
     echo ":::"
     echo "::: Commands:"
     echo ":::  [none]               Interactive mode"
     echo ":::  nopass               Create a client without a password"
     echo ":::  -d,--days            Expire the certificate after specified number of days (default: 1080)"
     echo ":::  -n,--name            Name for the Client (default: '"$(hostname)"')"
     echo ":::  -p,--password        Password for the Client (no default)"
+	echo ":::  -i,--iOS             Generate a certificate that leverages iOS keychain"
```
Comment (Giraffe1966) [inline scripts/makeOVPN.sh]: Perhaps include a link to https://openvpn.net/faq/how-do-i-use-a-client-certificate-and-private-key-from-the-ios-keychain/ so users know how to use the .ovpn12 file, and why this option increases security.

File: scripts/makeOVPN.sh [hunk 2]
```diff
@@ -235,8 +239,58 @@ if [ ! -f "${TA}" ]; then
...
+	# Copy the .ovpn profile to the home directory for convenient remote access
+		
+	printf "========================================================\n"
+	printf "Generating an .ovpn12 file for use with iOS devices\n"
+	printf "Please remember the export password\n"
+	printf "as you will need this import the certificate on your iOS device\n"
+	printf "========================================================\n"
+	openssl pkcs12 -passin pass:$PASSWD -passin pass:$PASSWD -export -in issued/${NAME}${CRT} -inkey private/${NAME}${KEY} -certfile ${CA} -name ${NAME} -out /home/$INSTALL_USER/ovpns/$NAME.ovpn12
+	chown "$INSTALL_USER" "/home/$INSTALL_USER/ovpns/$NAME.ovpn12"
+	chmod o-r "/home/$INSTALL_USER/ovpns/$NAME.ovpn12"
```
Comment (Giraffe1966) [inline scripts/makeOVPN.sh]:
```suggestion
	chmod 600 "/home/$INSTALL_USER/ovpns/$NAME.ovpn12"
```
`chmod 600` only allows read+write for owner (which is probably what most people would want), while `chmod o-r` might still allow a user who is not the owner or in the owning group to write to the file, if this is run on a user's machine with weird default permissions (where random users can write to others' files by default).

File: scripts/makeOVPN.sh [hunk 3]
```diff
@@ -235,8 +239,58 @@ if [ ! -f "${TA}" ]; then
...
+	} > "${NAME}${FILEEXT}"
+	
+	# Copy the .ovpn profile to the home directory for convenient remote access
+		
+	printf "========================================================\n"
+	printf "Generating an .ovpn12 file for use with iOS devices\n"
+	printf "Please remember the export password\n"
+	printf "as you will need this import the certificate on your iOS device\n"
+	printf "========================================================\n"
+	openssl pkcs12 -passin pass:$PASSWD -passin pass:$PASSWD -export -in issued/${NAME}${CRT} -inkey private/${NAME}${KEY} -certfile ${CA} -name ${NAME} -out /home/$INSTALL_USER/ovpns/$NAME.ovpn12
```
Comment (Giraffe1966) [inline scripts/makeOVPN.sh]: I believe using `-passin pass:$PASSWD` twice is a typo, and, according to the manpage, using `-passin env:PASSWD` might be more secure.

File: scripts/makeOVPN.sh (hunk 2 above)
Comment (4s3ti) [reply scripts/makeOVPN.sh]: Yup agree with that one there. 
also for sake of consistency and verbosity numbers should always be used.

File: scripts/makeOVPN.sh (hunk 1 above)
Comment (4s3ti) [reply scripts/makeOVPN.sh]: wouldn't a link be too much of information there? 
Maybe we should put the link in the FAQ or wiki or whatsoever and get a simple reference to it?

**Comment by 4s3ti** [comment]
@IcedComputer please apply the changes and suggestions mentioned by @Giraffe1966 and ill merge your PR

**Comment by 4s3ti** [comment]
Thanks @IcedComputer , Merging.

**Pull request merged**